scripts/templates/create-package/EmptyPackageJson.mustache

```diff
@@ -52,7 +52,9 @@
     "tslib": "{{{tslib}}}"
   },
   "peerDependencies": {
-    "react": "{{{react}}}",
-    "react-dom": "{{{reactDom}}}"
+    "@types/react": "{{{typesReactPeerDep}}}",
+    "@types/react-dom": "{{{typesReactDomPeerDep}}}",
+    "react": "{{{reactPeerDep}}}",
```

Comment:
I don't think we really need to do the @types/* peer deps anymore. I think that was related to a rush update fix.

Reply:
Okay so should I revert them all now? Change was made in #9798 for reference. It actually seems like a reasonable change since if people want to use OUFR with TS, they'll need React typings.

Reply:
I don't think this is a rush-related fix, but rather a Fabric + TypeScript related fix that we should keep.

Reply:
@kenotron As explained in https://github.com/OfficeDev/office-ui-fabric-react/issues/9720#issuecomment-509841203, if your package imports stuff from another package, you need to declare that dependency in your package.json.  If you don't declare it, then complex installations /will/ malfunction.  

Rush users tend to encounter these problems more often, because Rush's audience tends to be running serious production monorepos with complex installations.  Also, Rush [supports validation checks](https://github.com/microsoft/rush-example/blob/52054cda874e44c0f2aaf03cd2ee62068f4fcdbd/rush.json#L40) to proactively detect some of these problems.  Nonetheless, the problem applies to every package manager:  No package manager can perform a correct side-by-side installation for dependencies that were undeclared.

Reply:
Ken pointed out an interesting case, what about our consumers who don't use TypeScript? It's not ideal to make them either depend on `@types/react` and `@types/react-dom` or have unmet peer dependencies. But I also see the value of enforcing the correct typing versions for those consumers who do use TypeScript.

Reply:
PNPM and Yarn support "optional peer dependencies" to address this concern.  NPM supposedly will support it soon.

Reply:
I take a page out of what other component libs do:
https://github.com/ant-design/ant-design/blob/master/package.json

Putting @types/* seem to be a good idea for TS consumers, but kind of strange to enforce devdeps as peer deps. This is a lot like if we were to put raw-loader and webpack in our peer deps.

I simply did what I said in #9720 and blew away the lock file and did a rush update and it fixed everything without having to force libs consumed in rush repos to have (potentially varying) @types/react ranges: https://github.com/kenotron/RushWithMultipleFabricVersions

Reply:
According to [this (possibly dated) comment](https://github.com/discordjs/discord.js/pull/3066#issuecomment-461913077), NPM will simply ignore the optional peer dependencies field, resulting in the phantom dependency behavior that more-or-less works for simple setups.  That might be okay, actually. Projects that use NPM and plain JavaScript are unlikely to care about this.

Reply:
My point is that #9720's solution ended up adding a lot of peerdeps in our repo - we're happy to help consumers use our libs more easily, but the solution could have been for the OP for that bug to simply redo their lock files. Somehow things got gummed up when he created that lock file. That said, #9720 OP's repo could have had the same issue with many other libs that take @types/react as a devdep. I really don't think it's good practice to have consumers ask all lib authors to put @types/react in peerdeps.

Reply:
>  but the solution could have been for the OP for that bug to simply redo their lock files. Somehow things got gummed up when he created that lock file.

This viewpoint -- that lockfiles get "gummed up" and the fix is to jigger with them until the problem goes away -- it only works for simple setups.  At a certain scale, installation problems _cannot be solved_ without understanding very clearly how the package.json system works.  The first step to enlightenment is realizing that all those successful jiggerings have built up within yourself an overwhelming conviction that **_NPM versioning must be fundamentally simple_**.  It must all boil down to a few intuitive principles.  This conviction is incorrect.  It's flat wrong.  But it's very compelling when based on a body of experience where jiggerings worked great.

The OP already provided [a great repo](https://github.com/HarryGifford/RushWithMultipleFabricVersions).  I'd encourage you to look at it.  But if you're short on time, just consider for a moment how NPM is supposed to install a combined **node_modules** folder where Fabric 7 gets React 16, and Fabric 5 gets React 15.  Somehow NPM is supposed to guess which `@types/react` version goes in which folder, when the Fabric **package.json** doesn't even declare that it wants `@types/react` at all.  No amount of jiggering is going to sort that out.

> Another thought - I think folks using rush but NOT using TS but wishes to use the strictPeerDeps option will be forced to install the @types/react if they use OUFR. Not sure this is really desired outcome here.

See my note above about optional peers.

Reply:
> I simply did what I said in #9720 and blew away the lock file and did a rush update and it fixed everything without having to force libs consumed in rush repos to have (potentially varying) @types/react ranges: https://github.com/kenotron/RushWithMultipleFabricVersions

I just saw this comment -- lemme take a look at it.

Reply:
The `rush update` fixed everything because it inadvertently upgraded to `office-ui-fabric-react` 5.135.3 [with the peer dependencies](https://github.com/kenotron/RushWithMultipleFabricVersions/blob/bf011b2a7ec869dd2ece44bff51dfe3eed027c4b/common/config/rush/pnpm-lock.yaml#L2006) added.  😁

Reply:
I had a version of it locked to that version... will update my fork to illustrate. Still builds fine...

Reply:
This fix works because although you locked the direct dependency on `office-ui-fabric-react` to ["5.113.1"](https://github.com/kenotron/RushWithMultipleFabricVersions/blob/b48ed90c6a4244b0833e26fe70da1890e4770635/common/config/rush/pnpm-lock.yaml#L3082), the indirect dependencies still got upgraded. For example, `@uifabric/utilities` was solved as [5.34.3](https://github.com/kenotron/RushWithMultipleFabricVersions/blob/b48ed90c6a4244b0833e26fe70da1890e4770635/common/config/rush/pnpm-lock.yaml#L1998) which has the `@types/react` peer.  Because `office-ui-fabric-react` does not satisfy this peer, it gets determined by the parent project `test-15`.  This induces an implicit peer dependency for `office-ui-fabric-react@5.113.1` which is [recorded in the lockfile](https://github.com/kenotron/RushWithMultipleFabricVersions/blob/b48ed90c6a4244b0833e26fe70da1890e4770635/common/config/rush/pnpm-lock.yaml#L2004) like this:

```yaml
  /office-ui-fabric-react/5.113.1_ee63188b739fb4d6a436db584e63e7a3:
    dependencies:
      '@microsoft/load-themed-styles': 1.9.19
      '@uifabric/icons': 5.8.0_ee63188b739fb4d6a436db584e63e7a3
      '@uifabric/merge-styles': 5.17.1
      '@uifabric/styling': 5.37.0_ee63188b739fb4d6a436db584e63e7a3
      '@uifabric/utilities': 5.34.3_ee63188b739fb4d6a436db584e63e7a3
      prop-types: 15.7.2
      react: 15.6.2
      react-dom: 15.6.2_react@15.6.2
      tslib: 1.10.0
    dev: false
    peerDependencies:
      '@types/react': '*'
      '@types/react-dom': '*'
      react: ^0.14.9 || ^15.0.1-0 || ^16.0.0-0
      react-dom: ^0.14.9 || ^15.0.1-0 || ^16.0.0-0
```

If we really wanted to your theory, a reliable way would be to add some code like this to [common/config/rush/pnpmfile.js](https://github.com/kenotron/RushWithMultipleFabricVersions/blob/master/common/config/rush/pnpmfile.js):
```js
/**
 * This hook is invoked during installation before a package's dependencies
 * are selected.
 * The `packageJson` parameter is the deserialized package.json
 * contents for the package that is about to be installed.
 * The `context` parameter provides a log() function.
 * The return value is the updated object.
 */
function readPackage(packageJson, context) {
  if (packageJson.peerDependencies) {
    delete packageJson.peerDependencies['@types/react'];
    delete packageJson.peerDependencies['@types/react-dom'];
  }
  return packageJson;
}
```

This will revert the fix where we added `peerDependencies`, for all packages including indirect dependencies.  When I tried that and ran `rush update --full`, the build failure repros again.

Reply:
Would like to see why this lack of peerdeps would break the build there - the repo already specifies the actual devdep of @types.

Reply:
"Hints" isn't best term for these peer dependencies.  It's better to think of them as fixes for real bugs in package.json.  Side-by-side OUIFR version seems like an esoteric edge case, but it was just an easy example.  It's the tip of an iceberg.  Wherever package.json bugs are found, this same problem arises over and over and over again in a myriad of variations.  It's a real headache for developers.  As participants in the NPM ecosystem, we all need to work together and do our part to eliminate these bugs.

Although this investigation focused on PNPM, the same issue will repro easily with Yarn or NPM or even Yarn Plug'n'Play.  Their lockfiles are radically different, but the fix is always the same:  Declare the missing deps.

> So pnpm resolves these really specific dirs for its npm 2 style directories.

Heh, I suppose the deep node_modules paths are a bit reminiscent of "NPM 2". It's only a superficial similarity, though.  PNPM invented a [completely novel installation plan](https://pnpm.js.org/docs/en/symlinked-node-modules-structure.html) that solves the same important problems that Yarn recently tackled with their Plug'n'Play feature.  Whereas Plug'n'Play relies on redefining NodeJS's module resolution behavior, PNPM instead installs a node_modules folder with lots of nesting symlinks.  This weird symlinking strategy enabled PNPM to be backwards-compatible with existing NodeJS module resolution, which is how it was able to predate Plug'n'Play by a couple years.

Likening it to NPM 2 suggests that PNPM is old fashioned, but NPM 6 doesn't solve these problems at all.  (The specific problems I'm referring to are doppelgangers and phantom dependencies.)

> Conclusion - keep this in the create-package script. It's a good change.

Yay, thanks!  😊